src/webgpu/api/operation/render_pipeline/primitive_topology.spec.ts

```diff
@@ -0,0 +1,376 @@
+  | xx   xx xx   xx |     + Center of two triangle(s): {v2,v3,v4} and {v3,v4,v5}.
+  | x     x xxx xxx |
+  v1        v4 ---- v5
+`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { unreachable } from '../../../../common/framework/util/util.js';
+import { GPUTest } from '../../../gpu_test.js';
+
+const kRTSize: number = 32;
+const kColorFormat = 'rgba8unorm';
+const kValidPixelColor = new Uint8Array([0x00, 0xff, 0x00, 0xff]); // green
+const kInvalidPixelColor = new Uint8Array([0x00, 0x00, 0x00, 0x00]); // black
+
+class Point2D {
+  x: number;
+  y: number;
+  z: number;
+  w: number;
+
+  constructor(x: number, y: number) {
+    this.x = x;
+    this.y = y;
+    this.z = 0;
+    this.w = 1;
+  }
+
+  toNDC(): Point2D {
+    // NDC coordinate space is y-up, so we negate the y mapping.
+    // To ensure the resulting vertex in NDC will be placed at the center of the pixel, we
+    // must offset by the pixel coordinates or 0.5.
+    return new Point2D((2 * (this.x + 0.5)) / kRTSize - 1, (-2 * (this.y + 0.5)) / kRTSize + 1);
+  }
+
+  static getMidpoint(a: Point2D, b: Point2D) {
+    return new Point2D((a.x + b.x) / 2, (a.y + b.y) / 2);
+  }
+
+  static getCentroid(a: Point2D, b: Point2D, c: Point2D) {
+    return new Point2D((a.x + b.x + c.x) / 3, (a.y + b.y + c.y) / 3);
+  }
+}
+
+interface TestParams {
+  primitiveTopology: GPUPrimitiveTopology;
+}
+
+interface TestLocation {
+  location: Point2D;
+  color: Uint8Array;
+}
+
+const VertexLocations = [
+  new Point2D(8, 24), // v1
+  new Point2D(8, 8), // v2
+  new Point2D(16, 8), // v3
+  new Point2D(16, 24), // v4
+  new Point2D(24, 24), // v5
+  new Point2D(24, 8), // v6
+];
+
+class PrimitiveTopologyTest extends GPUTest {
+  makeAttachmentTexture(): GPUTexture {
+    return this.device.createTexture({
+      format: kColorFormat,
+      size: { width: kRTSize, height: kRTSize, depth: 1 },
+      sampleCount: 1,
+      usage: GPUTextureUsage.OUTPUT_ATTACHMENT | GPUTextureUsage.COPY_SRC,
+    });
+  }
+
+  makeVertexBuffer(vertexArray: Float32Array): GPUBuffer {
+    const [vertexBuffer, vertexMapping] = this.device.createBufferMapped({
+      size: vertexArray.byteLength,
+      usage: GPUBufferUsage.VERTEX,
+    });
+    new Float32Array(vertexMapping).set(vertexArray);
+    vertexBuffer.unmap();
+    return vertexBuffer;
+  }
+
+  static generateVertexBuffer(vertexLocations: Point2D[]): Float32Array {
+    const vertexCoords = new Float32Array(vertexLocations.length * 4);
+    for (let i = 0; i < vertexLocations.length; i++) {
+      const point = vertexLocations[i].toNDC();
+      vertexCoords[i * 4 + 0] = point.x;
+      vertexCoords[i * 4 + 1] = point.y;
+      vertexCoords[i * 4 + 2] = point.z;
+      vertexCoords[i * 4 + 3] = point.w;
+    }
+    return vertexCoords;
+  }
+
+  get params(): TestParams {
+    return super.params as TestParams;
+  }
+
+  static getPointTestLocations(expectedColor: Uint8Array): TestLocation[] {
```

Comment:
nit: these could become standalone functions (rather than static methods) just to minimize the amount of code inside the fixture class (would make the fixture more readable).

Reply:
Done.